File: examples/zig/if-without-else-multi-temp/StackTrackerReproV10min.runar.zig
const runar = @import("runar");

// StackTrackerReproV10min — minimal reproducer for issue #36 in Zig form.
//
// The first `if (... < outCount)` branch leaves `scriptLen`, `blobLen`, and
// `blob` as branch-private locals on the stack. Without the lowerIf branch
// reconciliation fix, post-ENDIF cleanup misindexed against `p` and the
// downstream OP_SPLIT aborted with "first stack item must be a non-negative
// number". Mirrors the TS fixture in examples/ts/if-without-else-multi-temp/.
pub const StackTrackerReproV10min = struct {
    pub const Contract = runar.SmartContract;

    pub fn init() StackTrackerReproV10min {
        return .{};
    }

    pub fn verifyMneeTxContainsBothOutputs(
        self: *const StackTrackerReproV10min,
        rawTx: runar.ByteString,
        expectedMneeOutputBytes: runar.ByteString,
        expectedExtraDataOutputBytes: runar.ByteString,
    ) void {
        _ = self;
        var p: i64 = 46;

        const outCount: i64 = runar.bin2num(runar.cat(runar.substr(rawTx, p, 1), runar.num2bin(0, 1)));
        runar.assert(outCount < 253);
        runar.assert(outCount <= 8);
        p = p + 1;

        var foundMnee: bool = false;
        var foundExtra: bool = false;

        if (0 < outCount) {
            const scriptLen: i64 = runar.bin2num(runar.cat(runar.substr(rawTx, p + 8, 1), runar.num2bin(0, 1)));
            runar.assert(scriptLen < 253);
            const blobLen: i64 = 8 + 1 + scriptLen;
            const blob: runar.ByteString = runar.substr(rawTx, p, blobLen);
            if (runar.bytesEq(blob, expectedMneeOutputBytes)) {
                foundMnee = true;
            }
            if (runar.bytesEq(blob, expectedExtraDataOutputBytes)) {
                foundExtra = true;
            }
            p = p + blobLen;
        }
        if (1 < outCount) {
            const scriptLen: i64 = runar.bin2num(runar.cat(runar.substr(rawTx, p + 8, 1), runar.num2bin(0, 1)));
            runar.assert(scriptLen < 253);
            const blobLen: i64 = 8 + 1 + scriptLen;
            const blob: runar.ByteString = runar.substr(rawTx, p, blobLen);
            if (runar.bytesEq(blob, expectedMneeOutputBytes)) {
                foundMnee = true;
            }
            if (runar.bytesEq(blob, expectedExtraDataOutputBytes)) {
                foundExtra = true;
            }
            p = p + blobLen;
        }

        runar.assert(foundMnee);
        runar.assert(foundExtra);
    }

    pub fn other(self: *const StackTrackerReproV10min, x: runar.ByteString) void {
        _ = self;
        runar.assert(runar.bytesEq(x, x));
    }
};

File: examples/ts/if-without-else-multi-temp/StackTrackerReproV10min.runar.ts
import { SmartContract, assert, substr, bin2num, num2bin, cat } from 'runar-lang';
import type { ByteString } from 'runar-lang';

class StackTrackerReproV10min extends SmartContract {
  constructor() { super(); }

  public verifyMneeTxContainsBothOutputs(
    rawTx: ByteString,
    expectedMneeOutputBytes: ByteString,
    expectedExtraDataOutputBytes: ByteString,
  ) {
    let p: bigint = 46n;

    const outCount: bigint = bin2num(cat(substr(rawTx, p, 1n), num2bin(0n, 1n)));
    assert(outCount < 0xfdn);
    assert(outCount <= 8n);
    p = p + 1n;

    let foundMnee: boolean = false;
    let foundExtra: boolean = false;

    if (0n < outCount) {
      const scriptLen: bigint = bin2num(cat(substr(rawTx, p + 8n, 1n), num2bin(0n, 1n)));
      assert(scriptLen < 0xfdn);
      const blobLen: bigint = 8n + 1n + scriptLen;
      const blob: ByteString = substr(rawTx, p, blobLen);
      if (blob === expectedMneeOutputBytes) { foundMnee = true; }
      if (blob === expectedExtraDataOutputBytes) { foundExtra = true; }
      p = p + blobLen;
    }
    if (1n < outCount) {
      const scriptLen: bigint = bin2num(cat(substr(rawTx, p + 8n, 1n), num2bin(0n, 1n)));
      assert(scriptLen < 0xfdn);
      const blobLen: bigint = 8n + 1n + scriptLen;
      const blob: ByteString = substr(rawTx, p, blobLen);
      if (blob === expectedMneeOutputBytes) { foundMnee = true; }
      if (blob === expectedExtraDataOutputBytes) { foundExtra = true; }
      p = p + blobLen;
    }

    assert(foundMnee);
    assert(foundExtra);
  }

  public other(x: ByteString) { assert(x === x); }
}

export default StackTrackerReproV10min;
